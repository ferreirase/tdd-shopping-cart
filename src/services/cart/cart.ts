import Dinero, { Dinero as DineroType } from 'dinero.js';
import { remove } from 'lodash';

const Money = Dinero;

Money.defaultCurrency = 'BRL';
Money.defaultPrecision = 2;

export type Product = {
  title: string;
  price: number;
};

export type DiscountCondition = {
  percentage?: number;
  minimum?: number;
  quantity?: number;
};

export type Item = {
  product: Product;
  quantity: number;
  condition?: DiscountCondition;
};

export function CalculatePercentageDiscount(
  amount: DineroType,
  item: Item,
): DineroType {
  if (
    item.condition?.percentage &&
    item.condition?.minimum &&
    item.quantity > item.condition?.minimum
  ) {
    return amount.percentage(item.condition.percentage);
  }

  return Money({ amount: 0 });
}

export function CalculateQuantityDiscount(
  amount: DineroType,
  item: Item,
): DineroType {
  const isEven = item.quantity % 2 === 0;

  if (
    item.condition &&
    item.condition.quantity &&
    item.quantity > item.condition.quantity
  ) {
    return amount.percentage(isEven ? 50 : 40);
  }

  return Money({ amount: 0 });
}

export class Cart {
  private items: Array<Item> = [];

  addItem(item: Item): void {
    const productAlreadyExists: Product | undefined = this.items.find(
      (currentItem: Item) => currentItem.product.title === item.product.title,
    )?.product;

    if (productAlreadyExists) {
      this.removeProduct(productAlreadyExists);
    }

    this.items.push(item);
  }

  removeProduct(product: Product): void {
    remove(this.items, { product });
  }

  getTotal(): number {
    return this.items
      .reduce((acc: DineroType, item: Item) => {
        const amount: DineroType = Money({
          amount: item.quantity * item.product.price,
        });

        const discount: DineroType = item.condition?.percentage
          ? CalculatePercentageDiscount(amount, item)
          : CalculateQuantityDiscount(amount, item);

        return acc.add(amount).subtract(discount);
      }, Money({ amount: 0 }))
      .getAmount();
  }

  summary(): { total: number; items: Array<Item> } {
    const total: number = this.getTotal();
    const items: Array<Item> = this.items;

    return {
      total,
      items,
    };
  }

  checkout(): ReturnType<typeof this.summary> {
    this.items = [];

    return this.summary();
  }
}

export { Money };
