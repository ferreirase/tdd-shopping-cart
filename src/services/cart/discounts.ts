import Dinero, { Dinero as DineroType } from 'dinero.js';
import { Item } from './cart';

export type DiscountCondition = {
  percentage?: number;
  minimum?: number;
  quantity?: number;
};

const Money = Dinero;

Money.defaultCurrency = 'BRL';
Money.defaultPrecision = 2;

export function CalculatePercentageDiscount(
  amount: DineroType,
  {
    condition,
    quantity,
  }: Omit<Item, 'condition'> & {
    condition?: DiscountCondition;
  },
): DineroType {
  if (
    condition?.percentage &&
    condition?.minimum &&
    quantity > condition?.minimum
  ) {
    return amount.percentage(condition.percentage);
  }

  return Money({ amount: 0 });
}

export function CalculateQuantityDiscount(
  amount: DineroType,
  {
    condition,
    quantity,
  }: Omit<Item, 'condition'> & {
    condition?: DiscountCondition;
  },
): DineroType {
  const isEven = quantity % 2 === 0;

  if (condition && condition.quantity && quantity > condition.quantity) {
    return amount.percentage(isEven ? 50 : 40);
  }

  return Money({ amount: 0 });
}

export function CalculateHigherDiscount(
  amount: DineroType,
  item: Omit<Item, 'condition'> & {
    condition: Array<DiscountCondition>;
  },
): DineroType {
  const { condition } = item;

  if (condition?.length === 0) {
    return Money({ amount: 0 });
  }

  const [discount] = condition
    .map((cond: DiscountCondition) => {
      return cond?.quantity
        ? CalculateQuantityDiscount(amount, {
            ...item,
            condition: cond,
          }).getAmount()
        : CalculatePercentageDiscount(amount, {
            ...item,
            condition: cond,
          }).getAmount();
    })
    .sort((a, b) => Number(b) - Number(a));

  return Money({ amount: discount as number });
}
