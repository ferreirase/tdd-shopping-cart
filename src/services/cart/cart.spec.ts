import { Cart, DiscountCondition, Item, Product } from './cart';

describe('Cart', () => {
  let cart: Cart;

  const product: Product = {
    title: 'Nike Jordan Masc',
    price: 400,
  };

  const product2: Product = {
    title: 'Nike Jordan Fem',
    price: 300,
  };

  beforeEach(() => {
    cart = new Cart();
  });

  describe('addItem()', () => {
    it('should ensure no more than one product exists at a time', () => {
      cart.addItem({
        product,
        quantity: 1,
      });

      cart.addItem({
        product,
        quantity: 2,
      });

      expect(cart.getTotal()).toEqual(800);
    });
  });

  describe('getTotal()', () => {
    it('should return 0 when getTotal is called first time', () => {
      expect(cart.getTotal()).toBe(0);
    });

    it('should multiply quantity and price and receive the total amount', () => {
      const item: Item = {
        product,
        quantity: 2,
      };

      cart.addItem(item);

      expect(cart.getTotal()).toEqual(item.quantity * item.product.price);
    });

    it('should update total when a product gets included and then removeItemd', () => {
      cart.addItem({
        product,
        quantity: 2,
      });

      cart.addItem({
        product: product2,
        quantity: 1,
      });

      cart.removeProduct(product);

      expect(cart.getTotal()).toEqual(300);
    });
  });

  describe('summary()', () => {
    it('should return the total and the list of items when summary() is called', () => {
      cart.addItem({
        product,
        quantity: 5,
      });

      cart.addItem({
        product: product2,
        quantity: 3,
      });

      expect(cart.summary()).toMatchSnapshot();
      expect(cart.getTotal()).toBeGreaterThan(0);
    });
  });

  describe('checkout()', () => {
    it('should return an object with the total and the list of items', () => {
      cart.addItem({
        product,
        quantity: 5,
      });

      cart.addItem({
        product: product2,
        quantity: 3,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });

    it('should reset the cart when checkout() is called', () => {
      cart.addItem({
        product: product2,
        quantity: 3,
      });

      cart.checkout();

      expect(cart.getTotal()).toEqual(0);
    });
  });

  describe('removeItem()', () => {
    it('should removeItem an product succesfully', () => {
      cart.addItem({
        product,
        quantity: 2,
      });

      cart.removeProduct(product);

      expect(cart.getTotal()).toBe(0);
    });
  });

  describe('special conditions', () => {
    it('should apply percentage discount when quantity above minimum is passed', () => {
      const condition = {
        percentage: 5,
        minimum: 3,
      };

      const item = {
        product,
        condition,
        quantity: 4,
      };

      cart.addItem(item);

      expect(cart.getTotal()).toBe(
        item.product.price * item.quantity -
          (item.product.price * item.quantity * condition.percentage) / 100,
      );
    });

    it('should NOT apply percentage discount when item quantity is below or equals condition minimum', () => {
      const condition: DiscountCondition = {
        minimum: 3,
        percentage: 10,
      };

      const item = {
        product,
        condition,
        quantity: 2,
      };

      cart.addItem(item);

      expect(cart.getTotal()).toBe(item.product.price * item.quantity);
    });

    it('should NOT apply quantity discount when item quantity is below or equals condition quantity', () => {
      const condition: DiscountCondition = {
        quantity: 3,
      };

      const item = {
        product,
        condition,
        quantity: 2,
      };

      cart.addItem(item);

      expect(cart.getTotal()).toBe(item.product.price * item.quantity);
    });

    it('should apply quantity discount for even quantities', () => {
      const condition = {
        quantity: 2,
      };

      const item = {
        product,
        condition,
        quantity: 4,
      };

      cart.addItem(item);

      expect(cart.getTotal()).toBe(
        (item.product.price * item.quantity) / condition.quantity,
      );
    });

    it('should apply quantity discount for odd quantities', () => {
      const condition = {
        quantity: 2,
      };

      const item = {
        product,
        condition,
        quantity: 5,
      };

      cart.addItem(item);

      expect(cart.getTotal()).toBe(
        item.product.price * item.quantity -
          item.product.price * item.quantity * 0.4,
      );
    });

    it('should apply the higher discount when 2 or more discount conditions is given', () => {
      const firstCondition: DiscountCondition = {
        quantity: 2,
      };

      const secondCondition: DiscountCondition = {
        percentage: 20,
        minimum: 2,
      };

      const item: Item = {
        product,
        condition: [firstCondition, secondCondition],
        quantity: 3,
      };

      cart.addItem(item);

      expect(cart.getTotal()).toBe(720);
    });

    it('should not apply discount when condition array is empty', () => {
      const item: Item = {
        product,
        condition: [],
        quantity: 3,
      };

      cart.addItem(item);

      expect(cart.getTotal()).toBe(item.product.price * item.quantity);
    });
  });
});
