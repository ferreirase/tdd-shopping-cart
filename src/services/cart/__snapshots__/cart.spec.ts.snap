// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total and the list of items 1`] = `
{
  "items": [],
  "total": 0,
}
`;

exports[`Cart summary() should return the total and the list of items when summary() is called 1`] = `
{
  "items": [
    {
      "product": {
        "price": 400,
        "title": "Nike Jordan Masc",
      },
      "quantity": 5,
    },
    {
      "product": {
        "price": 300,
        "title": "Nike Jordan Fem",
      },
      "quantity": 3,
    },
  ],
  "total": 2900,
}
`;
